### PR TITLE
Add render devices

### PR DESCRIPTION
**Pull request by est77**

### CODE REVIEW

File: src/appleseed/renderer/device/irenderdevice.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
+//
```
Comment (dictoon) [inline src/appleseed/renderer/device/irenderdevice.h]: Start file with a blank line.

File: src/appleseed/renderer/device/irendercontext.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
+//
```
Comment (dictoon) [inline src/appleseed/renderer/device/irendercontext.h]: Start file with a blank line.

File: src/appleseed/renderer/device/renderdevicebase.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+class RenderDeviceBase
+  : public IRenderDevice
+{
+  protected:
+    RenderDeviceBase(
+        Project&            project,
+        const ParamArray&   params);
+
+    ~RenderDeviceBase() override = default;
```
Comment (dictoon) [inline src/appleseed/renderer/device/renderdevicebase.h]: Can be removed.

File: src/appleseed/renderer/device/renderdevicebase.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    const Project& get_project() const;
+    Project& get_project();
+
+    const ParamArray& get_params() const;
+
+    bool is_progressive_render() const;
+
+    IRendererController* get_frame_renderer_controller() override;
```
Comment (dictoon) [inline src/appleseed/renderer/device/renderdevicebase.h]: Forward-declare `IRendererController`.

File: src/appleseed/renderer/device/cpu/cpurenderdevice.h [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#pragma once
+
+// appleseed.renderer headers.
+#include "renderer/device/renderdevicebase.h"
+#include "renderer/device/cpu/cpurendercontext.h"
```
Comment (dictoon) [inline src/appleseed/renderer/device/cpu/cpurenderdevice.h]: Order alphabetically.

File: src/appleseed/renderer/device/cpu/cpurenderdevice.h [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    bool initialize(
+        TextureStore&                   texture_store,
+        const foundation::SearchPaths&  resource_search_paths,
+        ITileCallbackFactory*           tile_callback_factory,
+        foundation::IAbortSwitch&       abort_switch) override;
+
+    bool build_or_update_scene() override;
+
+    bool load_checkpoint(Frame& frame, const size_t pass_count) override;
```
Comment (dictoon) [inline src/appleseed/renderer/device/cpu/cpurenderdevice.h]: Include `<cstddef>`.

Comment (dictoon) [reply src/appleseed/renderer/device/cpu/cpurenderdevice.h]: Many types would need to be forward-declared according to IWYU rules, even though they are provided by the base class' header file.

File: src/appleseed/renderer/kernel/rendering/masterrenderer.cpp [hunk 7]
```diff
@@ -231,91 +172,26 @@ struct MasterRenderer::Impl
...
         }
-
-        // Initialize the shader compiler, if the OSL headers are found.
-        if (m_resource_search_paths.exist("stdosl.h"))
+        else
         {
-            const APIString stdosl_path = m_resource_search_paths.qualify("stdosl.h");
-            RENDERER_LOG_INFO("found OSL headers in %s", stdosl_path.c_str());
-            m_osl_compiler = ShaderCompilerFactory::create(stdosl_path.c_str());
+            RENDERER_LOG_ERROR("Unknown renderer device %s.", device_name);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/masterrenderer.cpp]: `"unknown render device: %s."`

File: src/appleseed/renderer/kernel/rendering/masterrenderer.cpp [hunk 8]
```diff
@@ -465,7 +341,7 @@ struct MasterRenderer::Impl
         return input_binder.get_error_count() == 0;
     }
 
-    // Initialize rendering components and render a frame.
+    // Initialize renderer device and render a frame.
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/masterrenderer.cpp]: *render device*

File: src/appleseed/renderer/device/irenderdevice.h [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#pragma once
+
+// appleseed.renderer headers.
+#include "renderer/kernel/rendering/irenderercontroller.h"
```
[the remark this answers is not in the document]
Comment (est77) [reply src/appleseed/renderer/device/irenderdevice.h]: I can't. The render_frame method returns a IRendererController::Status

File: src/appleseed/renderer/kernel/rendering/masterrenderer.cpp [hunk 10]
```diff
@@ -498,22 +363,24 @@ struct MasterRenderer::Impl
             return renderer_controller.get_status();
         }
 
-        // Create renderer components.
-        RendererComponents components(
-            m_project,
-            m_params,
-            m_tile_callback_factory,
+        // Initialize the render device.
+        bool success = m_render_device->initialize(
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/masterrenderer.cpp]: Make const.

File: src/appleseed/renderer/device/cpu/cpurenderdevice.h (hunk 6 above)
Comment (est77) [reply src/appleseed/renderer/device/cpu/cpurenderdevice.h]: For types used in the public interface of the class, I think we can assume that they will be included or forward declared by base classes, no?

Comment (dictoon) [reply src/appleseed/renderer/device/cpu/cpurenderdevice.h]: Yes.

**Pull request merged**